diff --git a/pkg/app/client/client.go b/pkg/app/client/client.go
index 09a371a..809972f 100644
--- a/pkg/app/client/client.go
+++ b/pkg/app/client/client.go
@@ -491,6 +491,7 @@ func (c *Client) do(ctx context.Context, req *protocol.Request, resp *protocol.R
 		return fmt.Errorf("unsupported protocol %q. http and https are supported", scheme)
 	}
 	host := uri.Host()
+	sni := string(req.Header.Host())
 	startCleaner := false
 
 	c.mLock.Lock()
@@ -500,7 +501,7 @@ func (c *Client) do(ctx context.Context, req *protocol.Request, resp *protocol.R
 		m = c.ms
 	}
 
-	h := string(host)
+	h := string(sni)
 	hc := m[h]
 	if hc == nil {
 		if c.clientFactory == nil {
@@ -508,8 +509,9 @@ func (c *Client) do(ctx context.Context, req *protocol.Request, resp *protocol.R
 			c.clientFactory = factory.NewClientFactory(newHttp1OptionFromClient(c))
 		}
 		hc, _ = c.clientFactory.NewHostClient()
+		hc.SetClientSni(sni)
 		hc.SetDynamicConfig(&client.DynamicConfig{
-			Addr:     utils.AddMissingPort(h, isTLS),
+			Addr:     utils.AddMissingPort(string(host), isTLS),
 			ProxyURI: proxyURI,
 			IsTLS:    isTLS,
 		})
diff --git a/pkg/protocol/client/client.go b/pkg/protocol/client/client.go
index e0ed457..748f8c7 100644
--- a/pkg/protocol/client/client.go
+++ b/pkg/protocol/client/client.go
@@ -70,6 +70,7 @@ type HostClient interface {
 	CloseIdleConnections()
 	ShouldRemove() bool
 	ConnectionCount() int
+	SetClientSni(clientSni string)
 }
 
 type Doer interface {
diff --git a/pkg/protocol/http1/client.go b/pkg/protocol/http1/client.go
index c7b0a25..c428169 100644
--- a/pkg/protocol/http1/client.go
+++ b/pkg/protocol/http1/client.go
@@ -122,6 +122,7 @@ type HostClient struct {
 
 	tlsConfigMap     map[string]*tls.Config
 	tlsConfigMapLock sync.Mutex
+	ClientHelloSni   string
 
 	pendingRequests int32
 
@@ -811,6 +812,10 @@ func (c *HostClient) SetMaxConns(newMaxConns int) {
 	c.connsLock.Unlock()
 }
 
+func (c *HostClient) SetClientSni(clientSni string) {
+	c.ClientHelloSni = clientSni
+}
+
 func (c *HostClient) acquireConn(dialTimeout time.Duration) (cc *clientConn, inPool bool, err error) {
 	createConn := false
 	startCleaner := false
@@ -1185,6 +1190,11 @@ func (c *HostClient) cachedTLSConfig(addr string) *tls.Config {
 	cfg := c.tlsConfigMap[cfgAddr]
 	if cfg == nil {
 		cfg = newClientTLSConfig(c.TLSConfig, cfgAddr)
+
+		if c.ClientHelloSni != "" {
+			cfg.ServerName = c.ClientHelloSni
+		}
+
 		c.tlsConfigMap[cfgAddr] = cfg
 	}
 	c.tlsConfigMapLock.Unlock()
